File: test-suite-kotlin/src/test/kotlin/io/micronaut/kafka/docs/seek/ops/ProductListener.kt
package io.micronaut.kafka.docs.seek.ops

import io.micronaut.configuration.kafka.annotation.KafkaListener
import io.micronaut.configuration.kafka.annotation.OffsetReset
import io.micronaut.configuration.kafka.annotation.Topic
import io.micronaut.configuration.kafka.seek.KafkaSeekOperation
import io.micronaut.configuration.kafka.seek.KafkaSeekOperations
import io.micronaut.context.annotation.Property
import io.micronaut.context.annotation.Requires
import io.micronaut.kafka.docs.Product
import org.apache.kafka.common.TopicPartition

@KafkaListener(
    groupId = "kafka-seek-operations-group",
    offsetReset = OffsetReset.EARLIEST,
    properties = [Property(name = "max.poll.records", value = "1")]
)
@Requires(property = "spec.name", value = "KafkaSeekOperationsTest")
class ProductListener constructor(config: ProductListenerConfiguration) {

    var processed: MutableList<Product> = mutableListOf()

    @Topic("amazing-products")
    fun receive(product: Product, ops: KafkaSeekOperations) { // <1>
        processed.add(product)
        ops.defer(KafkaSeekOperation.seekToBeginning(TopicPartition("amazing-products", 0))) // <2>
    }
}

File: test-suite-python/src/test/python/micronaut/kafka/docs/consumer/topics/ProductListener.py
import logging
from typing import Annotated

from micronaut.configuration.kafka.annotation import KafkaKey, KafkaListener, Topic
from micronaut.context.annotation import Requires

LOG = logging.getLogger(__name__)


@Requires(property="spec.name", value="TopicsProductListenerTest")
@KafkaListener
class ProductListener:

    # tag::multiTopics[]
    @Topic(["fun-products", "awesome-products"])
    def receive_multi_topics(self, brand: Annotated[str, KafkaKey], name: str) -> None:
        LOG.info("Got Product - %s by %s", name, brand)
    # end::multiTopics[]

    # tag::patternTopics[]
    @Topic(patterns="products-\\w+")
    def receive_pattern_topics(self, brand: Annotated[str, KafkaKey], name: str) -> None:
        LOG.info("Got Product - %s by %s", name, brand)
    # end::patternTopics[]
